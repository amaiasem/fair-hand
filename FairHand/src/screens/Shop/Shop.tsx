import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'
import { StyleSheet, Text, View, Image, Linking, FlatList } from 'react-native'
import { COLOR, SIZES } from '../../../constants'
import { Ionicons, AntDesign } from '@expo/vector-icons'
import renderHeader from '../../Components/header/Header'
import { ScrollView, TouchableOpacity } from 'react-native-gesture-handler'
import { getReviewsByShopName } from '../../redux/actions/fairHandActionCreators'
import { bindActionCreators } from 'redux'
import Review from '../../Interfaces/reviewInterface'

const styles = StyleSheet.create({
  container: {
    marginTop: 30,
    flex: 1
  },
  shopDetail: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'center',
    width: SIZES.width
  },
  containerCover: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 220,
    width: SIZES.width
  },
  goBackButton: {
    position: 'absolute',
    width: SIZES.width,
    top: 0,
    padding: 10,
    zIndex: 1
  },
  icon: {
    fontSize: 40,
    color: COLOR.white
  },
  coverImage: {
    width: '100%',
    height: 220,
    resizeMode: 'cover'
  },
  containerLogo: {
    position: 'absolute',
    width: 120,
    height: 120,
    backgroundColor: COLOR.white,
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 60
  },
  logoImage: {
    width: 100,
    height: 100,
    resizeMode: 'contain',
    borderRadius: 120
  },
  shopInfo: {
    width: SIZES.width,
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    padding: 20
  },
  title: {
    fontSize: SIZES.h1,
    fontWeight: '700',
    marginBottom: 10,
    color: COLOR.grey
  },
  info: {
    fontSize: SIZES.p16,
    color: COLOR.black
  },
  web: {
    fontSize: SIZES.p16,
    color: COLOR.orange,
    fontStyle: 'italic'
  },
  containerButtons: {
    position: 'absolute',
    flexDirection: 'row',
    top: 0,
    right: 0,
    margin: 20
  },
  phoneHeart: {
    color: COLOR.black,
    marginLeft: 20
  },
  newInContainer: {
    marginLeft: 20,
    height: 280
  },
  newInTitle: {
    fontSize: SIZES.h2,
    marginRight: 20,
    fontWeight: '700',
    color: COLOR.black,
    borderBottomWidth: 1,
    borderBottomColor: COLOR.black
  },
  newInItem: {
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    width: 160,
    height: 235,
    marginTop: 10
  },
  newInImage: {
    height: 150,
    width: 150,
    resizeMode: 'cover',
    borderRadius: SIZES.cardRadius
  },
  newInInfo: {
    marginTop: 5,
    width: 150,
    flex: 1,
    justifyContent: 'space-between',
    alignItems: 'flex-start'
  },
  productName: {
    fontSize: SIZES.p12,
    fontWeight: '700'
  },
  price: {
    fontSize: SIZES.p12,
    color: COLOR.grey
  },
  newInbutton: {
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 5,
    width: 120,
    height: 25,
    backgroundColor: COLOR.orange,
    borderRadius: SIZES.buttonRadius
  },
  buttonText: {
    color: COLOR.white,
    fontSize: SIZES.p12
  },
  reviewsContainer: {
    marginTop: 15,
    marginBottom: 20,
    marginLeft: 20,
    marginRight: 20,
    width: '90%'
  },
  reviewsTitle: {
    fontSize: SIZES.h2,
    fontWeight: '700',
    color: COLOR.black,
    borderBottomWidth: 1,
    borderBottomColor: COLOR.black,
    marginBottom: 20
  },
  addReview: {
    position: 'absolute',
    right: 0,
    top: 0
  },
  reviewItem: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'flex-start'
  },
  reviewImage: {
    height: 40,
    width: 40,
    marginTop: 5,
    borderRadius: SIZES.buttonRadius
  },
  reviewInfo: {
    marginLeft: 15,
    marginBottom: 10,
    width: '85%'
  },
  userNameReview: {
    fontWeight: '700',
    fontSize: SIZES.p16
  },
  review: {
    fontSize: SIZES.p16
  }
})

const Shop = ({ reviews, action, route, navigation }:
  {reviews: Review[], action: any, route: any, navigation: any}) => {
  const { item }: any = route.params
  const [shop] = useState(item)

  useEffect(() => {
    action.getReviewsByShopName(item?.shopName)
  }, [])

  const renderItem = ({ item }: any) => (
    <View style={styles.newInItem}>
        <Image
          style = {styles.newInImage}
          source={{ uri: item.productImage }}
        ></Image>
        <View style={styles.newInInfo}>
          <View>
            <Text style={styles.productName}>{item.productName}</Text>
            {item.price ? <Text style={styles.price}>{item.price} â‚¬</Text> : <Text></Text>}
          </View>
          <TouchableOpacity
          testID='product-website'
          onPress= {() => { Linking.openURL(item.url) }}
          style={styles.newInbutton}>
            <Text style={styles.buttonText}>See on the website</Text>
          </TouchableOpacity>
        </View>
    </View>
  )

  return (
  <View style = {styles.container}>
    {renderHeader()}
    <ScrollView>
    <View style={styles.shopDetail}>
      <View style={styles.containerCover}>
        <View style={styles.goBackButton}>
            <Ionicons
              style={styles.icon}
              name="chevron-back-circle-outline"
              size={24}
              color="black"
              onPress={() => navigation.goBack()}
              testID='go-back'
              />
        </View>
        <Image
          style={styles.coverImage}
          source={{ uri: shop?.coverImage }}
          ></Image>
        <View style={styles.containerLogo}>
          <Image
          style={styles.logoImage}
          source={{ uri: shop?.logoImage }}
          ></Image>
        </View>
      </View>
      <View style={styles.shopInfo}>
        <Text style={styles.title}>{shop?.shopName}</Text>
        <Text style={styles.info}>{shop?.address}</Text>
        <Text style={styles.info}>{shop?.schedule}</Text>
        <TouchableOpacity
        testID= 'shop-website'
        onPress={() => Linking.openURL(`https://${shop?.website}`)}>
        <Text style={styles.web}>{shop?.website}</Text>
        </TouchableOpacity>
        <View style={styles.containerButtons}>
          <TouchableOpacity
          testID='shop-phone'
          onPress={() => Linking.openURL(`tel:${shop?.phone}`)}>
            <Ionicons style={styles.phoneHeart} name="call-outline" size={25} color="black" />
          </TouchableOpacity>
          <TouchableOpacity>
            <AntDesign style={styles.phoneHeart} name="hearto" size={25} color="black" />
          </TouchableOpacity>
        </View>
      </View>
      <View style={styles.newInContainer}>
        <Text style={styles.newInTitle}>New in</Text>
        <FlatList
          horizontal={true}
          showsHorizontalScrollIndicator={false}
          data={shop?.NewIn}
          renderItem={renderItem}
          keyExtractor={item => item._id}
          testID='list-products'
          ></FlatList>
      </View>
      <View style={styles.reviewsContainer}>
        <Text style={styles.reviewsTitle}>Reviews</Text>
        <View style={styles.addReview}>
          <AntDesign name="edit" size={24} color="black"
          testID='navigate-review'
          onPress={() => navigation.navigate('AddReview', { item })}/>
        </View>
        {
        reviews[0]
          ? reviews.map((review: Review, index: number) => (
              <View key={index} style={styles.reviewItem}>
              <Image style={styles.reviewImage} source={{ uri: review.image }}></Image>
              <View style={styles.reviewInfo}>
                <Text style={styles.userNameReview}>{review.userName}</Text>
                <Text style={styles.review}>{review.review}</Text>
              </View>
            </View>
          ))
          : <Text>There are no reviews</Text>
        }
      </View>
    </View>
    </ScrollView>
  </View>
  )
}

function mapStateToProps (state:any) {
  return {
    reviews: state.reviewReducer.reviews
  }
}

function mapDispatchToProps (dispatch: any) {
  return {
    action: bindActionCreators({ getReviewsByShopName }, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Shop)
