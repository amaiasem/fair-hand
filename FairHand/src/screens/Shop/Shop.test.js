import React from 'react'
import Shop from './Shop.tsx'
import renderer from 'react-test-renderer'

describe('Given a component Shop', () => {
  test('Then should have two children', () => {
    const shop = renderer.create(<Shop/>).toJSON()
    expect(shop.children.length).toBe(2)
    expect(shop).toMatchSnapshot()
  })

  test('Then should render correctlt', () => {
    const shop = renderer.create(<Shop/>).toJSON()
    expect(shop).toMatchSnapshot()
  })
})
