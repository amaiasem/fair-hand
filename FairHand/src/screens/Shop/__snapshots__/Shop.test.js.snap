// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a component Shop It should match the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Memo(Connect(Shop))
    navigation={
      Object {
        "goBack": [MockFunction],
        "navigate": [MockFunction],
      }
    }
    route={
      Object {
        "params": Object {
          "item": Object {
            "NewIn": Array [
              Object {
                "price": 119.9,
                "productImage": "https://cdn.shopify.com/s/files/1/0043/0804/1831/products/2059_image_4_5_front_2000x2500_crop_center.jpg?v=1614824553",
                "productName": "INKA GOLD PANTS",
                "url": "https://bravafabrics.com/products/inka-gold-pants",
              },
              Object {
                "price": 99.9,
                "productImage": "https://cdn.shopify.com/s/files/1/0043/0804/1831/products/2106_image_4_5_front_b95091e0-8724-4968-ba43-e1bfa00fe416_2000x2500_crop_center.jpg?v=1614789159",
                "productName": "MAI MAJOLICA DRESS",
                "url": "https://bravafabrics.com/products/mai-majolica-dress",
              },
              Object {
                "price": 74.9,
                "productImage": "https://cdn.shopify.com/s/files/1/0043/0804/1831/products/2006_image_4_5_front_28251a9e-1eb1-4b43-a8d8-b271966def02_2000x2500_crop_center.jpg?v=1614793112",
                "productName": "ALL NDEBELE SHIRT",
                "url": "https://bravafabrics.com/products/all-ndebele-shirt",
              },
            ],
            "_id": "343junx3dr2u",
            "address": "Parlament, 25 08015 Barcelona",
            "coverImage": "https://cdn.shopify.com/s/files/1/0043/0804/1831/files/cartela-web-EN_fdb23f61-d5dc-49e0-98f3-71fdb8298b3e_2000x.jpg?v=1614686668",
            "latlong": Object {
              "lat": 41.37644,
              "long": 2.16235,
            },
            "logoImage": "https://cdn.domestika.org/c_fill,dpr_auto,h_256,t_base_params.format_jpg,w_256/v1465550784/avatars/000/244/363/244363-original.png?1465550784",
            "phone": 931277197,
            "schedule": "Monday to Saturday 11h - 21h",
            "shopName": "Brava Fabrics",
            "type": Array [
              "clothes",
            ],
            "website": "https://bravafabrics.com/",
          },
        },
      }
    }
  />
</Provider>
`;
